File: services/backend-laravel/tests/Unit/RelationshipRuntimeLocalFakePayloadConsumerTest.php
<?php

namespace Tests\Unit;

use Tests\TestCase;

class RelationshipRuntimeLocalFakePayloadConsumerTest extends TestCase
{
    private function fixture(): array
    {
        $path = base_path('tests/Fixtures/relationship_runtime/relationship_runtime_contract_skeleton.json');
        $this->assertFileExists($path);

        $raw = file_get_contents($path);
        $this->assertNotFalse($raw);

        $data = json_decode((string) $raw, true);
        $this->assertIsArray($data);

        return $data;
    }

    private function localConsumerRows(array $fixture): array
    {
        return array_map(function (array $state): array {
            $capabilities = (array) ($state['capabilities'] ?? []);
            $claims = (array) ($state['runtime_claims'] ?? []);

            return [
                'id' => (string) ($state['id'] ?? ''),
                'state' => (string) ($state['state'] ?? ''),
                'round_id' => (string) ($state['round_id'] ?? ''),
                'fallback_copy_key' => (string) ($state['fallback_copy_key'] ?? ''),
                'can_reveal' => (bool) ($capabilities['can_reveal'] ?? false),
                'can_open_conversation' => (bool) ($capabilities['can_open_conversation'] ?? false),
                'manual_send_only' => $this->manualSendOnly($state),
                'local_feedback_only' => $this->localFeedbackOnly($state),
                'candidate_summary_present' => array_key_exists('candidate_summary', $state) && is_array($state['candidate_summary']),
                'candidate_present' => array_key_exists('candidate', $state) && is_array($state['candidate']),
                'candidate_source_labels' => (array) data_get($state, 'candidate.source_labels', []),
                'explanation_blocks_present' => ! empty($state['explanation_blocks']) && is_array($state['explanation_blocks']),
                'read_model_only' => (bool) data_get($state, 'conversation.read_model_only', false),
                'can_send' => (bool) data_get($state, 'conversation.can_send', false),
                'runtime_write_claimed' => in_array(true, array_values($claims), true),
                'forbidden_capability_enabled' => $this->hasForbiddenCapability($capabilities),
                'reason_code' => $state['reason_code'] ?? null,
            ];
        }, (array) ($fixture['states'] ?? []));
    }

    private function manualSendOnly(array $state): bool
    {
        $suggestions = (array) ($state['opener_suggestions'] ?? []);

        foreach ($suggestions as $suggestion) {
            $suggestion = (array) $suggestion;
            if (
                ($suggestion['manual_send_allowed'] ?? false) === true
                && ($suggestion['auto_send_allowed'] ?? true) === false
            ) {
                return true;
            }
        }

        return false;
    }

    private function localFeedbackOnly(array $state): bool
    {
        $feedback = (array) ($state['feedback'] ?? []);

        return ($feedback['local_only'] ?? false) === true
            || (
                ($feedback['can_record_local'] ?? false) === true
                && ($feedback['can_record_remote'] ?? true) === false
            );
    }

    private function hasForbiddenCapability(array $capabilities): bool
    {
        foreach ([
            'can_auto_send',
            'can_upload_media',
            'can_persist_feedback_remote',
            'can_push_notification',
        ] as $field) {
            if (($capabilities[$field] ?? false) === true) {
                return true;
            }
        }

        return false;
    }

    public function test_local_fake_payload_consumer_projects_all_contract_states(): void
    {
        $rows = $this->localConsumerRows($this->fixture());

        $this->assertCount(14, $rows);
        $this->assertSame([
            'not_started',
            'preparing',
            'waiting',
            'reveal_available',
            'revealed',
            'opener_prepared',
            'conversation_opened',
            'feedback_pending',
            'feedback_recorded_local',
            'feedback_recorded_remote_future',
            'next_round_pending',
            'no_candidate',
            'paused_or_unavailable',
            'auth_session_failure',
        ], array_column($rows, 'state'));

        foreach ($rows as $row) {
            $this->assertNotSame('', $row['id']);
            $this->assertNotSame('', $row['round_id']);
            $this->assertStringStartsWith('relationship.', $row['fallback_copy_key']);
            $this->assertFalse($row['runtime_write_claimed'], "state {$row['state']} must stay read-only");
            $this->assertFalse($row['forbidden_capability_enabled'], "state {$row['state']} enabled a forbidden capability");
        }
    }

    public function test_local_fake_payload_consumer_preserves_manual_send_and_read_model_boundaries(): void
    {
        $rows = collect($this->localConsumerRows($this->fixture()))->keyBy('state');

        $this->assertTrue($rows['reveal_available']['can_reveal']);
        $this->assertTrue($rows['reveal_available']['candidate_summary_present']);
        $this->assertFalse($rows['reveal_available']['candidate_present']);

        $this->assertTrue($rows['revealed']['candidate_present']);
        $this->assertTrue($rows['revealed']['explanation_blocks_present']);
        $this->assertContains('test_only', $rows['revealed']['candidate_source_labels']);
        $this->assertContains('not_runtime_evidence', $rows['revealed']['candidate_source_labels']);

        $this->assertTrue($rows['opener_prepared']['manual_send_only']);
        $this->assertTrue($rows['opener_prepared']['can_open_conversation']);
        $this->assertFalse($rows['opener_prepared']['runtime_write_claimed']);

        $this->assertTrue($rows['conversation_opened']['read_model_only']);
        $this->assertTrue($rows['conversation_opened']['can_open_conversation']);
        $this->assertFalse($rows['conversation_opened']['can_send']);
        $this->assertFalse($rows['conversation_opened']['manual_send_only']);
    }

    public function test_local_fake_payload_consumer_keeps_fallback_and_future_write_states_non_runtime(): void
    {
        $fixture = $this->fixture();
        $rows = collect($this->localConsumerRows($fixture))->keyBy('state');

        $this->assertTrue($rows['feedback_pending']['local_feedback_only']);
        $this->assertTrue($rows['feedback_recorded_local']['local_feedback_only']);
        $this->assertFalse($rows['feedback_recorded_remote_future']['runtime_write_claimed']);

        $this->assertFalse($rows['no_candidate']['candidate_present']);
        $this->assertSame('no_safe_candidate', $rows['no_candidate']['reason_code']);
        $this->assertSame('service_unavailable', $rows['paused_or_unavailable']['reason_code']);
        $this->assertSame('session_required', $rows['auth_session_failure']['reason_code']);

        $placeholders = (array) ($fixture['future_write_contract_placeholders'] ?? []);
        $mediaFuture = (array) ($placeholders['media_upload_future'] ?? []);
        $notificationPlaceholder = (array) ($placeholders['notification_return_placeholder'] ?? []);

        $this->assertFalse((bool) ($mediaFuture['runtime_write_allowed'] ?? true));
        $this->assertFalse((bool) ($notificationPlaceholder['read_state_write_allowed'] ?? true));
        $this->assertFalse((bool) ($notificationPlaceholder['push_delivery_claimed'] ?? true));
    }
}

File: apps/flutter_elitesync_module/test/features/match/data/relationship_runtime_local_fake_payload_consumer_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

class LocalRelationshipRuntimeView {
  const LocalRelationshipRuntimeView({
    required this.state,
    required this.roundId,
    required this.fallbackCopyKey,
    required this.canReveal,
    required this.canOpenConversation,
    required this.candidateSummaryPresent,
    required this.candidatePresent,
    required this.candidateSourceLabels,
    required this.explanationBlocksPresent,
    required this.manualSendOnly,
    required this.localFeedbackOnly,
    required this.readModelOnly,
    required this.canSend,
    required this.runtimeWriteClaimed,
    required this.forbiddenCapabilityEnabled,
    this.reasonCode,
  });

  final String state;
  final String roundId;
  final String fallbackCopyKey;
  final bool canReveal;
  final bool canOpenConversation;
  final bool candidateSummaryPresent;
  final bool candidatePresent;
  final List<String> candidateSourceLabels;
  final bool explanationBlocksPresent;
  final bool manualSendOnly;
  final bool localFeedbackOnly;
  final bool readModelOnly;
  final bool canSend;
  final bool runtimeWriteClaimed;
  final bool forbiddenCapabilityEnabled;
  final String? reasonCode;
}

void main() {
  Map<String, dynamic> loadFixture() {
    final file = File(
      '../../services/backend-laravel/tests/Fixtures/relationship_runtime/relationship_runtime_contract_skeleton.json',
    );
    expect(file.existsSync(), isTrue);
    return jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
  }

  LocalRelationshipRuntimeView consumeState(Map<String, dynamic> state) {
    final capabilities = state['capabilities'] as Map<String, dynamic>;
    final claims = state['runtime_claims'] as Map<String, dynamic>;
    final openerSuggestions =
        (state['opener_suggestions'] as List<dynamic>? ?? const <dynamic>[])
            .cast<Map<String, dynamic>>();
    final feedback =
        state['feedback'] as Map<String, dynamic>? ?? const <String, dynamic>{};
    final conversation =
        state['conversation'] as Map<String, dynamic>? ??
        const <String, dynamic>{};
    final candidate =
        state['candidate'] as Map<String, dynamic>? ??
        const <String, dynamic>{};

    final manualSendOnly = openerSuggestions.any(
      (item) =>
          item['manual_send_allowed'] == true &&
          item['auto_send_allowed'] == false,
    );
    final localFeedbackOnly =
        feedback['local_only'] == true ||
        (feedback['can_record_local'] == true &&
            feedback['can_record_remote'] == false);
    final forbiddenCapabilityEnabled = <String>[
      'can_auto_send',
      'can_upload_media',
      'can_persist_feedback_remote',
      'can_push_notification',
    ].any((key) => capabilities[key] == true);

    return LocalRelationshipRuntimeView(
      state: state['state'] as String,
      roundId: state['round_id'] as String,
      fallbackCopyKey: state['fallback_copy_key'] as String,
      canReveal: capabilities['can_reveal'] == true,
      canOpenConversation: capabilities['can_open_conversation'] == true,
      candidateSummaryPresent:
          state['candidate_summary'] is Map<String, dynamic>,
      candidatePresent: state['candidate'] is Map<String, dynamic>,
      candidateSourceLabels:
          (candidate['source_labels'] as List<dynamic>? ?? const <dynamic>[])
              .cast<String>(),
      explanationBlocksPresent:
          (state['explanation_blocks'] as List<dynamic>? ?? const <dynamic>[])
              .isNotEmpty,
      manualSendOnly: manualSendOnly,
      localFeedbackOnly: localFeedbackOnly,
      readModelOnly: conversation['read_model_only'] == true,
      canSend: conversation['can_send'] == true,
      runtimeWriteClaimed: claims.values.any((value) => value == true),
      forbiddenCapabilityEnabled: forbiddenCapabilityEnabled,
      reasonCode: state['reason_code'] as String?,
    );
  }

  List<LocalRelationshipRuntimeView> consumeFixture() {
    final fixture = loadFixture();
    return (fixture['states'] as List<dynamic>)
        .cast<Map<String, dynamic>>()
        .map(consumeState)
        .toList(growable: false);
  }

  test('local fake payload consumer projects every contract state', () {
    final views = consumeFixture();

    expect(views.map((view) => view.state), <String>[
      'not_started',
      'preparing',
      'waiting',
      'reveal_available',
      'revealed',
      'opener_prepared',
      'conversation_opened',
      'feedback_pending',
      'feedback_recorded_local',
      'feedback_recorded_remote_future',
      'next_round_pending',
      'no_candidate',
      'paused_or_unavailable',
      'auth_session_failure',
    ]);
    expect(
      views.map((view) => view.roundId),
      everyElement(startsWith('fixture-round-')),
    );
    expect(
      views.map((view) => view.fallbackCopyKey),
      everyElement(startsWith('relationship.')),
    );
    expect(
      views.map((view) => view.runtimeWriteClaimed),
      everyElement(isFalse),
    );
    expect(
      views.map((view) => view.forbiddenCapabilityEnabled),
      everyElement(isFalse),
    );
  });

  test(
    'local fake payload consumer preserves manual-send and read-model bounds',
    () {
      final views = {for (final view in consumeFixture()) view.state: view};

      expect(views['reveal_available']!.canReveal, isTrue);
      expect(views['reveal_available']!.candidateSummaryPresent, isTrue);
      expect(views['reveal_available']!.candidatePresent, isFalse);

      expect(views['revealed']!.candidatePresent, isTrue);
      expect(views['revealed']!.explanationBlocksPresent, isTrue);
      expect(views['revealed']!.candidateSourceLabels, contains('test_only'));
      expect(
        views['revealed']!.candidateSourceLabels,
        contains('not_runtime_evidence'),
      );

      expect(views['opener_prepared']!.manualSendOnly, isTrue);
      expect(views['opener_prepared']!.canOpenConversation, isTrue);
      expect(views['opener_prepared']!.runtimeWriteClaimed, isFalse);

      expect(views['conversation_opened']!.readModelOnly, isTrue);
      expect(views['conversation_opened']!.canOpenConversation, isTrue);
      expect(views['conversation_opened']!.canSend, isFalse);
      expect(views['conversation_opened']!.manualSendOnly, isFalse);
    },
  );

  test(
    'local fake payload consumer preserves fallback and future-gate bounds',
    () {
      final views = {for (final view in consumeFixture()) view.state: view};
      final placeholders =
          loadFixture()['future_write_contract_placeholders']
              as Map<String, dynamic>;
      final mediaFuture =
          placeholders['media_upload_future'] as Map<String, dynamic>;
      final notificationPlaceholder =
          placeholders['notification_return_placeholder']
              as Map<String, dynamic>;

      expect(views['feedback_pending']!.localFeedbackOnly, isTrue);
      expect(views['feedback_recorded_local']!.localFeedbackOnly, isTrue);
      expect(
        views['feedback_recorded_remote_future']!.runtimeWriteClaimed,
        isFalse,
      );

      expect(views['no_candidate']!.reasonCode, 'no_safe_candidate');
      expect(views['paused_or_unavailable']!.reasonCode, 'service_unavailable');
      expect(views['auth_session_failure']!.reasonCode, 'session_required');
      expect(mediaFuture['runtime_write_allowed'], isFalse);
      expect(notificationPlaceholder['read_state_write_allowed'], isFalse);
      expect(notificationPlaceholder['push_delivery_claimed'], isFalse);
    },
  );
}
